### PR TITLE
support cast/try_cast in prune with signed integer and decimal

### PR DESCRIPTION
**Pull request by liukun4515**
# Which issue does this PR close?

part of  https://github.com/apache/arrow-datafusion/issues/3414
Closes https://github.com/apache/arrow-datafusion/issues/3377

In order to merge this  https://github.com/apache/arrow-datafusion/pull/3396, we should support the cast/try_cast in the prune first.

cc @andygrove @alamb 

Just support cast from unsigned integer or decimal to unsigned integer or decimal.

 # Rationale for this change

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (andygrove) [inline datafusion/core/src/physical_optimizer/pruning.rs:1577]: I'm not sure I understand what is happening with the negated case. Could you add a comment here explaining this? Why is there `!=` in here rather than `<=` or `>=`?

**Review by alamb — COMMENTED** [review]
Thank you @liukun4515  -- this is great

I will look at this code in the morning as well, but I think we need to handle the case where casting won't preserve sortedness (aka UTF8 -> numeric), as described on https://github.com/apache/arrow-datafusion/issues/3377

**Comment by liukun4515** [comment]
> Thank you @liukun4515 -- this is great
> 
> I will look at this code in the morning as well, but I think we need to handle the case where casting won't preserve sortedness (aka UTF8 -> numeric), as described on #3377

I got your point.

We can add rule for cast/try_cast case.

For example, just support cast/try_cast from numeric to numeric data type, I think the data type of numeric has the order.

We can support other complex case in the future pr.

@alamb

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:1577]: @andygrove I don't change any logical about the list.
you can get this from https://github.com/apache/arrow-datafusion/pull/2282#discussion_r855486677

I get the meanings from the code context.
if the min_c1 and max_c1 is 1, all the value in the column or chunk must be 1 or null, so the `c1 not in (1)` can skip this column or chunk.

**Comment by liukun4515** [comment]
cc @alamb 
The cast/try_cast just support signed integer/decimal to signed integer/decimal.

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:495]: Could you please update the list of rules in the docstring of this function to include the new logic for casts?

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:441]: Very minor suggestion (can do this as a follow on too)

```suggestion
        let (column_expr, correct_operator, scalar_expr) = rewrite_expr_to_prunable(
            column_expr,
            correct_operator,
            scalar_expr,
            df_schema,
        )?;
```

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:518]: We could probably avoid the duplication of the error generation by moving the error return into `support_type_for_prune` like:

```suggestion
            verify_support_type_for_prune(&from_type, data_type)?;
            let (left, op, right) = rewrite_expr_to_prunable(expr, op, scalar_expr)?;
            Ok((cast(left, data_type.clone()), op, right))
```

This is not needed for correctness, just a code structure suggestion

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:582]: 👍  thank you for these comments

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:585]: I think it would be good to file a follow on ticket listing the other types (you did so in https://github.com/apache/arrow-datafusion/issues/3414#issuecomment-1242659478)

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:2104]: I think we should add some tests that exercise the pruning code with actual values too

For example, perhaps you can take `prune_int32_cast` from https://github.com/apache/arrow-datafusion/pull/3378?

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:2085]: 👍

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (alamb) [inline datafusion/core/src/physical_optimizer/pruning.rs:2079]: Rather than extending this test for other ops, I think we should add tests which actually run the pruning expressions on values (see my other comments).

Would it be helpful for me to write some more tests? I can do so tomorrow morning

**Comment by alamb** [comment]
To be explicit I think the code in this PR looks good to me, but I would like to see some actual tests with pruning expressions on data prior to merging it. I am happy to help write these tests too

**Comment by liukun4515** [comment]
> To be explicit I think the code in this PR looks good to me, but I would like to see some actual tests with pruning expressions on data prior to merging it. I am happy to help write these tests too

@alamb  Thanks for you help.
It's too later for me to add test today, but welcome to add test for this pr, I can help to review your test and other changes tomorrow.
I will never do any changes before merged your test or your suggestion.

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:2079]: I find i have added the test for executing with real data in the `prune_cast_column_scalar` function.

You can check them, it just include cast(col, int64) and try_cast(col,int64). If you think it is not enough, you can add more test for it or give some advices for the test.

Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:2079]: https://github.com/apache/arrow-datafusion/blob/4b1a4052d079c8149ad232a05e9965cceb05b5fc/datafusion/core/src/physical_optimizer/pruning.rs#L2011

https://github.com/apache/arrow-datafusion/blob/4b1a4052d079c8149ad232a05e9965cceb05b5fc/datafusion/core/src/physical_optimizer/pruning.rs#L2022

https://github.com/apache/arrow-datafusion/blob/4b1a4052d079c8149ad232a05e9965cceb05b5fc/datafusion/core/src/physical_optimizer/pruning.rs#L2028

FYI @alamb

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:441]: update manually

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:518]: good point

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs:585]: https://github.com/apache/arrow-datafusion/issues/3442

File: datafusion/core/src/physical_optimizer/pruning.rs
Comment (Dandandan) [inline datafusion/core/src/physical_optimizer/pruning.rs:521]: This comment should be removed I think?

Comment (liukun4515) [reply datafusion/core/src/physical_optimizer/pruning.rs]: Done
removed

**Review by alamb — APPROVED** [review]
I re-reviewed this @liukun4515  and I think it looks good to go. Thank you.

I wrote several more tests which I will put up in a follow on PR but they all pass and I don't think there is any reason to hold this PR.

**Comment by alamb** [comment]
Follow on PR is https://github.com/apache/arrow-datafusion/pull/3454

**Comment by liukun4515** [comment]
Thanks for your careful review. @alamb

**Pull request merged**
